### PR TITLE
(470) First cut of the Stock reports

### PR DESCRIPTION
**Pull request by parisiale**
Adding functionality for creating and exporting;
 - stock summary report;
 - stock movement report.

### CODE REVIEW

File: models/stock_export.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    date = fields.Date('Date')
+
+    # @todo: (api) is this decorator necessary? Idem below
+    @api.model
+    def __get_default_stock_location(self):
+        # @todo: (ale) without the parent, this will also include
+        # locations like 'TRW/Stock' (for which the name is actually
+        # 'Stock', ofc...); ensure we want to also specify it
+        return self.env['stock.location'].search([('name', '=', 'Stock'),
+                                                  ('location_id', '=', 'UPL')])
```
Comment (crwr45) [inline models/stock_export.py]: location_id should be matching an integer, shouldn't it? By my understanding this shouldn't work.

File: models/stock_export.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        Product = self.env['product.product']
+        Location = self.env['stock.location']
+        all_quants = Quant.search([('location_id', 'child_of', locations.ids)])
+        by_prod = lambda x: x.product_id.id
+        by_location = lambda x: x.location_id.id
+
+        get_pallets_by_loc = lambda quant_union: {
+            Location.browse(loc): Quant.union(*_quants).mapped('package_id')
+            for loc, _quants in groupby(quant_union.sorted(key=by_location),
+                                        key=by_location)}
```
Comment (crwr45) [inline models/stock_export.py]: Looks like this lambda is getting the package the quant is in, which is not necessarily a pallet. 
We are removing 'pallet' from the terminology and using 'package' and 'parent package'. In this case I think it should be 'package'.

File: models/stock_export.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+            quant_prods = Product.browse(prod)
+            quants_u = Quant.union(*_quants)
+            pallets_by_prod_by_loc[quant_prods] = get_pallets_by_loc(quants_u)
+            pallets_by_prod[quant_prods] = quants_u.mapped('package_id')
+
+        def get_prod_qty(prod, pallet):
+            contained_quants = Quant.search(
+                [('package_id', 'child_of', pallet.ids)])
+            prod_quants = contained_quants.filtered(
+                lambda x: x.product_id == prod)
```
Comment (crwr45) [inline models/stock_export.py]: The filter should be moved to be part of the initial search domain.

File: models/stock_export.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        '''
+        if not self.date:
+            raise ValidationError(_("Date not found."))
+
+        self.field_data = False
+        self.field_name = False
+
+        Move = self.env['stock.move']
+        picking_type_in = self.env.ref('stock.picking_type_in')
+        picking_type_out = self.env.ref('stock.picking_type_out')
```
Comment (crwr45) [inline models/stock_export.py]: These picking types need to be accessed via the logged in users current warehouse (Users.get_user_warehouse())

File: models/stock_export.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        out_moves = moves.filtered(lambda m: m.picking_type_id == picking_type_out)
+
+        _logger.info(_('Creating Excel file'))
+        file_name = 'warehouse_movemenet_{TIMESTAMP}.xls'.format(
+            TIMESTAMP=datetime.now())
+        wb = xlwt.Workbook()
+
+        for goods_moves, sheet_name in [(in_moves, "Goods In"),
+                                        (out_moves, "Goods Out")]:
+            sheet = wb.add_sheet(sheet_name)
```
Comment (crwr45) [inline models/stock_export.py]: The method above used to create a sheet should be used here.

File: models/stock_export.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+             'datas_fname': file_name})
+
+        _send_message_to_user(
+            self.env['res.users'],
+            subject="%s Ready" % doc_title,
+            body=_("%s %s is attached.") % (doc_title, file_name),
+            attachment=attachment)
+
+
+# @todo: (ale) consider moving this helper elsewhere
```
Comment (crwr45) [inline models/stock_export.py]: Definitely move it elsewhere. I would suggest either adding to to the res.users model, or the mail.message model.

**Review by crwr45 — COMMENTED** [review]
The references to 'pallet' should be genericised to 'package'.

The code structure is also not sitting right with me. There are two different reports exported by the same class, shown to the user as two different menu items. I think it would make sense to have the data extraction and file creation code on seperate models for the two reports.

File: models/stock_export.py (hunk 1 above)
Comment (parisiale) [reply models/stock_export.py]: Oh, mistake, I've tried this with `odoo-bin shell` by using the `name` attribute.

File: models/stock_export.py (hunk 4 above)
Comment (parisiale) [reply models/stock_export.py]: Thanks, will change.

File: models/stock_export.py (hunk 5 above)
Comment (parisiale) [reply models/stock_export.py]: Yeah, but the above one is an inner closure, around the `Workbook` instance, so, if you make it an actual method, you should add that to the args. Besides that, another reason why I did not refactor is that I'm no sure of the impact of adding methods to model classes (and btw, is it necessary to add the `@api.model` decorator?).

**Comment by parisiale** [comment]
@crwr45 thanks!

The code structure follows prior art.
I'm not sure I understand the benefit of what you're suggesting with the two separate models for data / file. The file-related logic (determine the file name and process a `xlwt` workbook) is tied to the fact we're using the excel format. And `xlwt` is something you use in the method that retrieves the data.

File: models/stock_export.py (hunk 1 above)
Comment (parisiale) [reply models/stock_export.py]: I think we can simply remove the parent check.

**Review by crwr45 — APPROVED** [review]

**Pull request merged**